Make ReactiveClientHeadersFactory replace headers instead of adding

michalszynkiewicz opened the pull request:
fixes #24054

michalszynkiewicz committed:
Make ReactiveClientHeadersFactory replace headers instead of adding

fixes #24054

michalszynkiewicz commented:
CC @Sgitario

geoand commented on extensions/resteasy-reactive/rest-client-reactive/runtime/src/main/java/io/quarkus/rest/client/reactive/runtime/MicroProfileRestClientRequestFilter.java:
```diff
@@ -72,6 +72,7 @@ public void filter(ClientRequestContext requestContext) {
                 reactiveRequestContext.suspend();
                 MultivaluedMap<String, String> outgoingHeaders = incomingHeaders;
                 reactiveClientHeadersFactory.getHeaders(incomingHeaders).subscribe().with(newHeaders -> {
+                    outgoingHeaders.clear();
```
Can you explain why this is needed? Prefereably in a source-code comment so it's persistent :)

michalszynkiewicz replied:
that's the way the factory in the MP spec works. It takes in headers that come from an incoming request and creates a new map of headers from it. It's basically about propagating selected headers.
I'll add a comment about it. But I started to have some doubts re how this code works with calls that do blocking stuff so will first do some experiments :)

Sgitario reviewed (commented):
Is this how is working the resteasy classic vendor of the reactive client header annotation?

If not, I'm not sure if this should be the right behavior. Mostly because of the scenario of using this annotation in sub-resources where the headers of the latest resource will take precedent.

michalszynkiewicz commented:
> Is this how is working the resteasy classic vendor of the reactive client header annotation?
> 
> If not, I'm not sure if this should be the right behavior. Mostly because of the scenario of using this annotation in sub-resources where the headers of the latest resource will take precedent.

the `incomingHeaders` are not the headers from previous annotations, etc but headers from a call that initiated this client call.
So, if e.g. curl calls our endpoint with a set of headers, then the endpoint uses the client to call another endpoint. `incomingHeadeders` will contain headers sent by curl.

michalszynkiewicz commented:
By the way I also added the header params that are defined by the client as a parameter to this method. This is how it works in the classic variant. 

I had to change how the suspend/resume is done - it wasn't working for async invocations. I added the timer in the test to check for it. This test was failing before the change.

geoand commented on extensions/resteasy-reactive/rest-client-reactive/runtime/src/main/java/io/quarkus/rest/client/reactive/ReactiveClientHeadersFactory.java:
```diff
@@ -10,7 +10,20 @@
+     *
+     * @see ClientHeadersFactory#update(MultivaluedMap, MultivaluedMap)
+     */
+    public abstract Uni<MultivaluedMap<String, String>> getHeaders(MultivaluedMap<String, String> incomingHeaders,
```
This is a breaking change, so we should add the appropriate label and ready an entry for the migration guide.

michalszynkiewicz commented:
https://github.com/quarkusio/quarkus/wiki/Migration-Guide-2.8#rest-client-reactive-reactiveclientheadersfactory-method-change - migration guide

michalszynkiewicz commented:
Without it the reactive factory is not much use for real blocking stuff so I'm wondering if even though the change is blocking, isn't it worth backporting it.

WDYT @geoand  @gsmet ?

geoand commented:
I'm +0
